gt/main.c

```diff
@@ -244,13 +250,94 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
 		"gt: receiving a packet with IP source address %s, and destination address %s, whose destination IP address is not the Grantor server itself.!\n",
 		src, dst);
 }
+static void
+gt_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+
+	if (pcall_again != NULL)
+		*pcall_again = true;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ * The fields @ref_cnt and @ip_addr are not used by GT blocks.
+ */
+static inline int
+gt_fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+	uint16_t inner_ip_ver, void *ip_dst,
+	struct gatekeeper_if *iface, int lcore_id)
+{
+	int ret;
+
+	eth_cache->stale = true;
+	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(inner_ip_ver);
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		ret = hold_arp(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
```

Comment:
That's true. In GK, we need to delete some Ethernet cache entries pointed by some FIB entries, so we need ``put_xx()`` function. But, in GT, we don't have a way to delete these entries. Maybe for IPv4, it has enough space for holding the entries for all the protected subnet, but for IPv6, we may need to find a way to ``replace`` old entries.